File: frontend/app/src/utils/nocobase.js
import api from '../api'
import logger from './logger'

const DEFAULT_PAGE_SIZE = 500

export function normalizeId(value) {
  if (value === null || value === undefined || value === '') return value
  if (typeof value === 'number') return value
  if (/^\d+$/.test(String(value))) return Number(value)
  return value
}

export function toFilter(filter) {
  return JSON.stringify(filter)
}

export function dedupe(values = []) {
  return [...new Set(values.filter((value) => value !== null && value !== undefined && value !== ''))]
}

function toArray(value) {
  if (Array.isArray(value)) return value
  if (typeof value === 'string' && value) return value.split(',').map((s) => s.trim())
  return undefined
}

function normalizeParams(params) {
  const result = { ...params }
  if (result.appends !== undefined) result.appends = toArray(result.appends)
  if (result.sort !== undefined) result.sort = toArray(result.sort)
  return result
}

export async function list(collection, params = {}) {
  const normalized = normalizeParams({ pageSize: DEFAULT_PAGE_SIZE, ...params })
  logger.debug(`list ${collection}`, normalized)
  const { data } = await api.get(`/${collection}:list`, { params: normalized })
  logger.debug(`list ${collection} → ${data.data?.length ?? 0} records`)
  return data.data || []
}

export async function get(collection, id, params = {}) {
  const normalized = normalizeParams({ filterByTk: id, ...params })
  logger.debug(`get ${collection}`, normalized)
  const { data } = await api.get(`/${collection}:get`, { params: normalized })
  logger.debug(`get ${collection} → id=${data.data?.id}`)
  return data.data
}

export async function create(collection, payload) {
  logger.debug(`create ${collection}`, payload)
  const { data } = await api.post(`/${collection}:create`, payload)
  logger.debug(`create ${collection} → id=${data.data?.id}`)
  return data.data
}

export async function update(collection, id, payload) {
  logger.debug(`update ${collection} id=${id}`, payload)
  const { data } = await api.post(`/${collection}:update?filterByTk=${id}`, payload)
  logger.debug(`update ${collection} id=${id} → ok`)
  return data.data
}

export async function destroy(collection, id) {
  logger.debug(`destroy ${collection} id=${id}`)
  try {
    await api.delete(`/${collection}:destroy?filterByTk=${id}`)
  } catch {
    await api.post(`/${collection}:destroy?filterByTk=${id}`)
  }
  logger.debug(`destroy ${collection} id=${id} → ok`)
}

File: backend/src/core/nocobase.py
"""
HTTP-клиент для взаимодействия с NocoBase REST API.

Предоставляет CRUD-операции над коллекциями: list, get, create, update,
destroy, set_relation.  Каждый вызов открывает отдельный ``httpx.AsyncClient``
(stateless), что упрощает конфигурацию и потокобезопасность.
"""

from __future__ import annotations

import json
from typing import Any

import httpx
from fastapi import HTTPException

from .config import settings
from .logging import get_logger

logger = get_logger(__name__)

# Максимальное количество записей, запрашиваемых за один раз
DEFAULT_PAGE_SIZE = 500


class NocoBaseClient:
    """Асинхронный клиент к NocoBase API."""

    def _normalize_params(self, params: dict[str, Any]) -> dict[str, Any]:
        """Привести параметры запроса к плоскому строковому виду."""
        normalized: dict[str, Any] = {}
        for key, value in params.items():
            if value is None:
                continue
            if isinstance(value, (list, tuple)):
                normalized[key] = ','.join(str(item) for item in value)
            else:
                normalized[key] = value
        return normalized

    async def _request(
        self,
        method: str,
        path: str,
        *,
        params: dict[str, Any] | None = None,
        json_payload: Any = None,
    ) -> httpx.Response:
        """Выполнить HTTP-запрос к NocoBase и вернуть ответ."""
        if not settings.api_key:
            raise HTTPException(status_code=500, detail='API_KEY is not configured')

        headers = {
            'Authorization': f'Bearer {settings.api_key}',
            'Content-Type': 'application/json',
        }

        logger.debug("NocoBase → %s %s  params=%s", method, path, params)

        async with httpx.AsyncClient(
            base_url=settings.nocobase_api_url,
            timeout=settings.nocobase_timeout,
            headers=headers,
        ) as client:
            response = await client.request(
                method,
                path,
                params=self._normalize_params(params or {}),
                json=json_payload,
            )

        if response.status_code >= 400:
            detail: Any = response.text
            try:
                detail = response.json()
            except Exception:
                pass
            logger.warning("NocoBase ← %s %s  status=%d", method, path, response.status_code)
            raise HTTPException(status_code=response.status_code, detail=detail)

        logger.debug("NocoBase ← %s %s  status=%d", method, path, response.status_code)
        return response

    async def list(
        self,
        collection: str,
        *,
        filter: dict[str, Any] | None = None,
        appends: str | list[str] | None = None,
        sort: str | list[str] | None = None,
        page_size: int = DEFAULT_PAGE_SIZE,
    ) -> list[dict[str, Any]]:
        """Получить список записей коллекции с фильтрацией и сортировкой."""
        response = await self._request(
            'GET',
            f'/{collection}:list',
            params={
                'pageSize': page_size,
                'filter': json.dumps(filter) if filter is not None else None,
                'appends': appends,
                'sort': sort,
            },
        )
        payload = response.json()
        return payload.get('data') or []

    async def get(
        self,
        collection: str,
        filter_by_tk: str | int,
        *,
        appends: str | list[str] | None = None,
    ) -> dict[str, Any]:
        """Получить одну запись коллекции по первичному ключу."""
        response = await self._request(
            'GET',
            f'/{collection}:get',
            params={'filterByTk': filter_by_tk, 'appends': appends},
        )
        return response.json().get('data') or {}

    async def create(self, collection: str, payload: dict[str, Any]) -> dict[str, Any]:
        """Создать запись в коллекции."""
        response = await self._request('POST', f'/{collection}:create', json_payload=payload)
        return response.json().get('data') or {}

    async def update(self, collection: str, record_id: str | int, payload: dict[str, Any]) -> dict[str, Any]:
        """Обновить запись коллекции по ID."""
        response = await self._request(
            'POST',
            f'/{collection}:update',
            params={'filterByTk': record_id},
            json_payload=payload,
        )
        return response.json().get('data') or {}

    async def destroy(self, collection: str, record_id: str | int) -> None:
        """Удалить запись из коллекции (пробует DELETE, при ошибке — POST)."""
        try:
            await self._request('DELETE', f'/{collection}:destroy', params={'filterByTk': record_id})
        except HTTPException:
            await self._request('POST', f'/{collection}:destroy', params={'filterByTk': record_id})

    async def set_relation(self, collection: str, record_id: str | int, relation: str, ids: list[str | int]) -> Any:
        """Установить связь many-to-many, пробуя несколько стратегий NocoBase API."""
        strategies = [
            lambda: self.update(collection, record_id, {relation: ids}),
            lambda: self.update(collection, record_id, {relation: [{'id': value} for value in ids]}),
            lambda: self._request('POST', f'/{collection}/{record_id}/{relation}:set', json_payload={'values': ids}),
            lambda: self._request('POST', f'/{collection}/{record_id}/{relation}:set', json_payload=ids),
        ]

        last_error: HTTPException | None = None
        for strategy in strategies:
            try:
                return await strategy()
            except HTTPException as error:
                last_error = error

        if last_error:
            raise last_error
        raise HTTPException(status_code=500, detail='Failed to update relation')


nocobase_client = NocoBaseClient()
